### PR TITLE
[ricos-editor] feat(essentials-models) command & ricos-content PoC

### PR DESCRIPTION
**Pull request by maimonav**

### CODE REVIEW

File: packages/ricos-editor/web/src/models/ricos-content.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import type { RefinedNode, RichTextNode, RichContentNodeWithContainerData } from 'ricos-content';
+import type { PluginContainerData, TextData } from 'ricos-schema';
+
+/**
+ * Represents Editor content node.
+ * Adds functionality on top of RefinedNode (which, on turn, is based on proto-derived Node)
+ *
+ * @template RefinedNodeType specific Refined Node
+ */
+export interface IRicosNode<RefinedNodeType extends RefinedNode> {
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/ricos-content.ts:10]: IRicosNode => RicosNode

File: packages/ricos-editor/web/src/models/ricos-content.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+// TODO: refactor name
+export interface IRicosNodeWithContainerData<DT>
+  extends IRicosNode<RichContentNodeWithContainerData> {
+  getContainerData(): PluginContainerData;
+  setContainerData(data: PluginContainerData): void;
+  getData(): DT;
+  setData(data: DT): void;
+}
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/ricos-content.ts:30]: Merge them into one, remove containerData & textData

export interface RicosContentNode<DT> extends RicosNode<RefinedNode>{
	getData
	setData
}

File: packages/ricos-editor/web/src/models/ricos-paragraph-node.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    this.node = cloneDeep(props.node);
+    this.selected = props.selected;
+  }
+
+  getTextData(): TextData[] {
+    return this.node.nodes.map(textNode => textNode.textData);
+  }
+
+  setTextData(data: TextData[]): void {
+    const nodes: TextNode[] = data.map(textData => generateTextNode(textData));
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/ricos-paragraph-node.tsx:31]: Create RicosTextNode as Domain entity & RicosTextNodeFactory & RicosTextNodesAgg and RicosParagraphNodesAgg will use it.

File: packages/ricos-editor/web/src/models/ricos-paragraph-node.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+});
+
+type Props = { node: ParagraphNode; selected: boolean };
+
+export class RicosParagraphNode implements IRicosTextNode<ParagraphData> {
+  node: ParagraphNode;
+
+  selected: boolean;
+
+  constructor(props: Props) {
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/ricos-paragraph-node.tsx:21]: props to parameters (node, selected)

File: packages/ricos-editor/web/src/models/ricos-content.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Represents Editor content node.
+ * Adds functionality on top of RefinedNode (which, on turn, is based on proto-derived Node)
+ *
+ * @template RefinedNodeType specific Refined Node
+ */
+export interface IRicosNode<RefinedNodeType extends RefinedNode> {
+  node: RefinedNodeType;
+  selected: boolean;
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/ricos-content.ts:12]: move selected to impl
add getSelected, setSelected

Comment (maimonav) [reply packages/ricos-editor/web/src/models/ricos-content.ts:12]: static of(paragraphNode) : RicosParagraphNode
private constructor

File: packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import type { IRicosNodes } from './ricos-content';
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx:1]: change to ts in all files

File: packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import { RicosParagraphNode } from './ricos-paragraph-node';
+import type { ParagraphNode } from 'ricos-content';
+
+// TODO: selected value?
+
+export class RicosParagraphNodes implements IRicosNodes<RicosParagraphNode> {
+  nodes: RicosParagraphNode[];
+
+  constructor(nodes: ParagraphNode[]) {
+    const paragraphNodes = nodes.map(node => new RicosParagraphNode({ node, selected: false }));
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx:11]: RicosParagraphNodeFactory

File: packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    this.nodes = paragraphNodes;
+  }
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx:13]: RicosParagraphNodes.of(nodes: ParagraphNode[]) : RicosParagraphNodes
private constructor(RicosParagraphNode[])

File: packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+  filter(predicate: (node: RicosParagraphNode) => boolean): RicosParagraphNodes {
+    this.nodes = this.nodes.filter(predicate);
+    return this;
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx:17]: const nodes = this.nodes.filter(predicate);
return new privateConstructor

Comment (maimonav) [reply packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx:17]: https://github.com/wix/ricos/blob/7f53795805b96956951ca70d1e55c4aea7f5a6e4/packages/tiptap-editor/web/src/models/Extensions.ts#L62-L70

File: packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    this.nodes = [...this.nodes, new RicosParagraphNode({ node, selected: false })];
+    return this;
+  }
+
+  update(node: RicosParagraphNode): RicosParagraphNodes {
+    throw new Error('Method not implemented.');
+  }
+
+  delete(node: RicosParagraphNode): RicosParagraphNodes {
+    throw new Error('Method not implemented.');
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/ricos-paragraph-nodes.tsx:30]: use filter

File: packages/ricos-editor/web/src/models/ricos-paragraph-node.tsx (hunk 3 above)
Comment (maimonav) [reply packages/ricos-editor/web/src/models/ricos-paragraph-node.tsx:31]: Changed to getTextNode & setTextNode

File: packages/ricos-editor/web/src/models/ricos-content.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+ */
+export interface RicosNode<RefinedNodeType extends RefinedNode> {
+  node: RefinedNodeType;
+  getSelected(): boolean;
+  setSelected(selected: boolean): void;
+  // style: NodeStyle;, TODO: need to address it (getStyle, setStyle)
+}
+
+export interface RicosContentNode<DT> extends RicosNode<RefinedNode> {
+  // might need to declare abstract class with static "of" method
```
Comment (lxgreen) [inline packages/ricos-editor/web/src/models/ricos-content.ts:17]: `RicosNode` is gonna extend `Pointed` interface (the one with `of` method). I'm on it.

File: packages/ricos-editor/web/src/models/ricos-node-factory.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
+import type { IRicosNodeFactory } from './ricos-content';
+import type { RefinedNode } from 'ricos-content';
+import { RicosParagraphNode } from './ricos-paragraph-node';
+import { Node_Type } from 'ricos-schema';
+import { RicosTextNode } from './ricos-text-node';
+
+export class RicosNodeFactory implements IRicosNodeFactory<RefinedNode> {
```
Comment (lxgreen) [inline packages/ricos-editor/web/src/models/ricos-node-factory.ts:7]: factory should be as stupid as possible:
```
// the generic factory parametrized by RefinedNode and Pointed derivatives.
// In this case, it is just to call  the `of` exposed by the K.
interface RicosNodeFactory<T extends RefinedNode, K extends Pointed> {
  create(node: T): K;
}


export class RicosParagraphNodeFactory implements RicosNodeFactory<ParagraphNode, RicosParagraphNode> {
  create(node: ParagraphNode) {
    return RicosParagraphNode.of(node);
}
```

File: packages/ricos-editor/web/src/models/ricos-paragraph-node.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
+import type { ParagraphData, TextData } from 'ricos-schema';
+import { Node_Type } from 'ricos-schema';
+import type { ParagraphNode, TextNode } from 'ricos-content';
+import { RicosNodeFactory } from './ricos-node-factory';
+import type { RicosContentNode } from './ricos-content';
+
+const generateTextNode = (textData: TextData): TextNode => ({
```
Comment (lxgreen) [inline packages/ricos-editor/web/src/models/ricos-paragraph-node.ts:7]: factory

File: packages/ricos-editor/web/src/models/ricos-paragraph-node.ts [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+  getData(): ParagraphData {
+    return this.node.paragraphData;
+  }
+
+  setData(data: ParagraphData): void {
+    this.node.paragraphData = data;
+  }
+
+  getTextNodes(): TextNode[] {
+    return this.node.nodes;
```
Comment (lxgreen) [inline packages/ricos-editor/web/src/models/ricos-paragraph-node.ts:45]: should expose aggregate of `RicosTextNode`, not the nodes

File: packages/ricos-editor/web/src/models/ricos-paragraph-node.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  setTextNodes(nodes: TextNode[]): void {
```
Comment (lxgreen) [inline packages/ricos-editor/web/src/models/ricos-paragraph-node.ts:48]: this should be handled by `RicosTextNodes` (`RicosTextNode` aggregate)

File: packages/ricos-editor/web/src/models/ricos-paragraph-nodes.ts [hunk 16]
```diff
@@ -0,0 +1,40 @@
+import type { RicosNodes } from './ricos-content';
+import type { RicosParagraphNode } from './ricos-paragraph-node';
+import { RicosNodeFactory } from './ricos-node-factory';
+import type { ParagraphNode } from 'ricos-content';
+
+export class RicosParagraphNodes implements RicosNodes<RicosParagraphNode> {
+  nodes: RicosParagraphNode[];
+
+  static of(nodes: ParagraphNode[]) {
+    const factory = new RicosNodeFactory();
```
Comment (lxgreen) [inline packages/ricos-editor/web/src/models/ricos-paragraph-nodes.ts:10]: should be injected via parameters rather instantiated. this way, it is easy to replace the implementation

File: packages/ricos-editor/web/src/models/ricos-paragraph-nodes.ts [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+    this.nodes = nodes;
+  }
+
+  filter(predicate: (node: RicosParagraphNode) => boolean): RicosParagraphNodes {
+    const nodes = this.nodes.filter(predicate);
+    return new RicosParagraphNodes(nodes);
+  }
+
+  insert(node: ParagraphNode): RicosParagraphNodes {
+    const factory = new RicosNodeFactory();
```
Comment (lxgreen) [inline packages/ricos-editor/web/src/models/ricos-paragraph-nodes.ts:25]: the factory should be a class private property, every method should use it

File: packages/ricos-editor/web/src/models/ricos-text-nodes.ts [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  insert(node: TextNode): RicosTextNodes {
+    const factory = new RicosNodeFactory();
+    const ricosNode = factory.create(node) as RicosTextNode;
+    const nodes = [...this.nodes, ricosNode];
+    return new RicosTextNodes(nodes);
+  }
+
+  // how to impl update & delete? there is no id for text node
```
Comment (lxgreen) [inline packages/ricos-editor/web/src/models/ricos-text-nodes.ts:31]: let's talk about it today ;)

**Review by lxgreen — CHANGES_REQUESTED** [review]

File: packages/ricos-content/web/src/types/node-data-refined-types.ts [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+  Decoration_Type.LINK,
+  Decoration_Type.MENTION,
+  Decoration_Type.SPOILER,
+  Decoration_Type.UNDERLINE,
+];
+
+// TODO: validate specific data items?
+const isDecoration = (data: Decoration): data is Decoration =>
+  DECORATION_TYPES.includes(data.type) &&
+  (!!data.anchorData ||
```
Comment (maimonav) [inline packages/ricos-content/web/src/types/node-data-refined-types.ts:25]: Decoration might have no data

File: packages/ricos-editor/web/src/models/RicosParagraphNodes/ricos-paragraph-node.ts [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+  getData(): ParagraphData {
+    return this.node.paragraphData;
+  }
+
+  setData(data: ParagraphData): void {
+    this.node.paragraphData = data;
+  }
+
+  getRicosTextNodes(): RicosTextNodes {
+    return RicosTextNodes.of(this.node.nodes);
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/RicosParagraphNodes/ricos-paragraph-node.ts:37]: new => property

File: packages/ricos-editor/web/src/models/draft-ricos-node-repository.ts [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+  updateSelection(): RicosContent;
+}
+
+export class DraftRicosNodeRepository implements IDraftRicosNodeRepository {
+  _ricosContent: RicosContent;
+
+  editor: DraftEditor;
+
+  constructor(editorState: EditorState) {
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/draft-ricos-node-repository.ts:27]: See if needed

File: packages/ricos-editor/web/src/models/RicosTextNodes/ricos-text-nodes.ts [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+  private readonly descriptors: RicosNodeDescriptors;
+
+  private get nodes(): RicosTextNode[] {
+    return this.descriptors.getNodes() as RicosTextNode[];
+  }
+
+  static of(nodes: TextNode[]) {
+    const factory = new RicosParagraphTextNodesFactory();
+    const descriptorFactory = new RicosNodeDescriptorFactory(factory);
+    const ricosNodesDescriptors = nodes.map(node => descriptorFactory.create(node));
```
[the remark this answers is not in the document]
Comment (lxgreen) [reply packages/ricos-editor/web/src/models/RicosTextNodes/ricos-text-nodes.ts:26]: this way, we get factory from outside **not** via constructor, and we just pass it further via constructor.

File: packages/ricos-editor/web/src/models/RicosTextNodes/ricos-text-nodes.ts [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+    const descriptors = this.descriptors.filter(predicate);
+    return new RicosTextNodes(descriptors);
+  }
+
+  insert(node: TextNode): RicosTextNodes {
+    const factory = new RicosParagraphTextNodesFactory();
+    const descriptorFactory = new RicosNodeDescriptorFactory(factory);
+    const ricosNodeDescriptor = descriptorFactory.create(node);
+    const descriptors = this.descriptors.insert(ricosNodeDescriptor);
+    return new RicosTextNodes(descriptors);
```
Comment (lxgreen) [inline packages/ricos-editor/web/src/models/RicosTextNodes/ricos-text-nodes.ts:45]:
```suggestion
    // after you fix factory... 
    return new RicosTextNodes(this.descriptors.insert(this.factory.create(node)), this.factory);
```

**Review by lxgreen — CHANGES_REQUESTED** [review]

**Comment by lxgreen** [comment]
review summary:
- please move all implementations outside `models` (as name suggests, it is supposed to contain only models -- interfaces)
- see the comments regarding the factory injection and factory composition
- let's decide on naming conventions (private variable and interface prefixes)

File: packages/ricos-editor/web/src/content-modification/nodes/ricos-nodes.ts [hunk 24]
```diff
@@ -0,0 +1,63 @@
+import type { RefinedNode } from 'ricos-content';
+import type { RicosNode, RicosNodes } from '../../models/ricos-content';
+import type { DescriptorManager, IRicosNodeDescriptors } from '../../models/ricos-descriptor';
+import { RicosNodeDescriptors } from './ricos-node-descriptors';
+
+export class RicosTrackedNodes<TNode extends RefinedNode, TRicosNode extends RicosNode<TNode>>
```
Comment (maimonav) [inline packages/ricos-editor/web/src/content-modification/nodes/ricos-nodes.ts:6]: Rename TrackedNodes?

File: packages/ricos-editor/web/src/content-modification/nodes/ricos-nodes.ts [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+  implements RicosNodes<TNode>, DescriptorManager {
+  private readonly descriptors: IRicosNodeDescriptors;
+
+  private mapper: (nodes: TNode) => TRicosNode;
+
+  private get nodes(): TRicosNode[] {
+    return this.descriptors.getNodes() as TRicosNode[];
+  }
+
+  static withMapper<TNode extends RefinedNode, TRicosNode extends RicosNode<TNode>>(
```
Comment (maimonav) [inline packages/ricos-editor/web/src/content-modification/nodes/ricos-nodes.ts:17]: Rename Mapper?

File: packages/ricos-editor/web/src/models/command.ts [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+ *  }
+ *
+ *  ```
+ */
+export interface Commands {
+  [name: CommandDescriptor['name']]: <Args>(
+    args: Args
+  ) => ReturnType<CommandDescriptor<Args>['execute']>;
+}
+
```
Comment (maimonav) [inline packages/ricos-editor/web/src/models/command.ts:76]: Create `ricos-commands` & `wix-ricos-commands` packages.

File: packages/ricos-editor/web/src/content-modification/services/draft-content-repository.ts [hunk 27]
```diff
@@ -0,0 +1,33 @@
+import type { ParagraphNode } from 'ricos-content';
+import { fromDraft } from 'ricos-content/libs/fromDraft';
+import { toDraft } from 'ricos-content/libs/toDraft';
```
Comment (maimonav) [inline packages/ricos-editor/web/src/content-modification/services/draft-content-repository.ts:3]: @lxgreen The bundle increment is due to these two, we either dynamically import them or need to implement node-converter for paragraph & text.

Comment (lxgreen) [reply packages/ricos-editor/web/src/content-modification/services/draft-content-repository.ts:3]: > to implement node-converter for paragraph & text
i strongly suggest we do this. then, the code is clean and detached from legacy

**Comment by maimonav** [comment]
@lxgreen after dynamic loading of converters, the bundle size still impacted:
ricos-editor-no-modal: 883KB => 895KB
ricos-editor-with-plugins: 1095KB => 1109KB
ricos-editor: 883KB => 895KB
The change is about ~13KB, which I believe this is a result of usages of ricos-schema, but I'm not sure, this will need to be investigated.
Nevertheless, I think it's ok and that this change is not an issue.
WDYT?

**Pull request merged**